Add inference providers support for Hugging Face (#8258)
BerriAI/litellm: docs/my-website/docs/providers/huggingface.md

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Hugging Face
LiteLLM supports running inference across multiple services for models hosted on the Hugging Face Hub.

> - **Serverless Inference Providers** - Hugging Face offers an easy and unified access to serverless AI inference through multiple inference providers, like [Together AI](https://together.ai) and [Sambanova](https://sambanova.ai). This is the fastest way to integrate AI in your products with a maintenance-free and scalable solution. More details in the [Inference Providers blogpost](https://huggingface.co/blog/inference-providers).

Author: we don't have a proper inference providers docs, but once we do, I will update the documentation here!

- **Dedicated Inference Endpoints** - which is a product to easily deploy models to production. Inference is run by Hugging Face in a dedicated, fully managed infrastructure on a cloud provider of your choice. You can deploy your model on Hugging Face Inference Endpoints by following the steps [here](https://huggingface.co/docs/inference-endpoints/guides/create_endpoint).

## Supported Models

### Serverless Inference Providers
You can check available models for an inference provider by going to [huggingface.co/models](https://huggingface.co/models), clicking the "Other" filter tab, and selecting your desired provider:

> ![Filter models by Inference Provider](../../img/hf_filter_inference_providers.png)

Reviewer: yes, will be deployed soon (today).

And yes i agree with adding a link to the filtered view

> ![Filter models by Inference Provider](../../img/hf_filter_inference_providers.png)

Reviewer 2: and it's deployed: https://huggingface.co/models?inference_provider=together&sort=trending

### Dedicated Inference Endpoints
Refer to the [Inference Endpoints catalog](https://endpoints.huggingface.co/catalog) for a list of available models.

## Usage

<Tabs>
<TabItem value="serverless" label="Serverless Inference Providers">

### Authentication
With a single Hugging Face token, you can access inference through multiple providers. Your calls are routed through Hugging Face and the usage is billed directly to your Hugging Face account at the standard provider API rates.

Simply set the `HF_TOKEN` environment variable with your Hugging Face token, you can create one here: https://huggingface.co/settings/tokens.

```bash
export HF_TOKEN="hf_xxxxxx"
```
or alternatively, you can pass your Hugging Face token as a parameter:
```python
completion(..., api_key="hf_xxxxxx")
```

### Getting Started

To use a Hugging Face model, specify both the provider and model you want to use in the following format:
```
huggingface/<provider>/<model_id>
```
Where `<model_id>` is the Hugging Face model ID and `<provider>` is the inference provider.

> By default, if you don't specify a provider, LiteLLM will use the [HF Inference API](https://huggingface.co/docs/api-inference/en/index).

Reviewer 3: Maybe add an example of this below?

Reviewer: this is mostly for backward compat with the existing LiteLLM implementation, correct? cc @ishaan-jaff

Examples:

```python
# Run DeepSeek-R1 inference through Together AI
completion(model="huggingface/together/deepseek-ai/DeepSeek-R1",...)

# Run Qwen2.5-72B-Instruct inference through Sambanova
completion(model="huggingface/sambanova/Qwen/Qwen2.5-72B-Instruct",...)
```

> For Hugging Face models, LiteLLM uses Hugging Face's Python client [InferenceClient](https://huggingface.co/docs/huggingface_hub/package_reference/inference_client) under the hood. `InferenceClient` is a drop-in replacement for OpenAI client.

Reviewer 2: ```suggestion
For Hugging Face models, LiteLLM uses Hugging Face's Python client [InferenceClient](https://huggingface.co/docs/huggingface_hub/package_reference/inference_client) under the hood. `InferenceClient` is a drop-in replacement for OpenAI client with some extra features to handle providers.
```

<a target="_blank" href="https://colab.research.google.com/github/BerriAI/litellm/blob/main/cookbook/LiteLLM_HuggingFace.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

### Basic Completion
Here's an example of chat completion using the DeepSeek-R1 model through Together AI:

```python
import os
from litellm import completion

os.environ["HF_TOKEN"] = "hf_xxxxxx"

response = completion(
    model="huggingface/together/deepseek-ai/DeepSeek-R1",

>     messages=[{"content": "How many r's are in the word `strawberry`?", "role": "user"}]

Reviewer 4: nit: maybe put `"role"` first, it seems to be more usual.

)
print(response)
```

### Streaming
Now, let's see what a streaming request looks like.

```python
import os
from litellm import completion

os.environ["HF_TOKEN"] = "hf_xxxxxx"

response = completion(
    model="huggingface/together/deepseek-ai/DeepSeek-R1",
    messages=[{"content": "How many r's are in the word `strawberry`?", "role": "user"}],
    stream=True
)

for chunk in response:
    print(chunk)
```

### Image Input

> You can also pass images to the model.

Reviewer 2: ```suggestion
You can also pass images when the model supports it. Here is an example using [Llama-3.2-11B-Vision-Instruct](https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct) model through Sambanova.
```

```python
from litellm import completion

# Set your Hugging Face Token
os.environ["HF_TOKEN"] = "hf_xxxxxx"

messages=[
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "What's in this image?"},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                    }
                },
            ],
        }
    ]

response = completion(

>     model="huggingface/sambanova/Qwen/Qwen2.5-72B-Instruct", 

Reviewer 2: ```suggestion
    model="huggingface/sambanova/meta-llama/Llama-3.2-11B-Vision-Instruct", 
```

Let's use a VLM like https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct  (https://huggingface.co/Qwen/Qwen2.5-72B-Instruct is not AFAIK)

    messages=messages,
)
print(response.choices[0])
```

> ### Function Calling
> You can extend the model's capabilities by giving them access to tools, here is an example with function calling using the Qwen2.5-72B-Instruct model through Sambanova.

Reviewer 2: ```suggestion
### Function Calling
You can extend the model's capabilities by giving them access to tools. Here is an example with function calling using [Qwen2.5-72B-Instruct](https://huggingface.co/Qwen/Qwen2.5-72B-Instruct) model through Sambanova.
```

from litellm import completion

# Set your Hugging Face Token
os.environ["HF_TOKEN"] = "hf_xxxxxx"

tools = [
  {
    "type": "function",
    "function": {
      "name": "get_current_weather",
      "description": "Get the current weather in a given location",
      "parameters": {
        "type": "object",
        "properties": {
          "location": {
            "type": "string",
            "description": "The city and state, e.g. San Francisco, CA",
          },
          "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
        },
        "required": ["location"],
      },
    }
  }
]
messages = [{"role": "user", "content": "What's the weather like in Boston today?"}]

response = completion(
    model="huggingface/sambanova/meta-llama/Llama-3.3-70B-Instruct", 
    messages=messages,
    tools=tools,
    tool_choice="auto"
)
print(response)
```

</TabItem>

<TabItem value="endpoints" label="Inference Endpoints">

<a target="_blank" href="https://colab.research.google.com/github/BerriAI/litellm/blob/main/cookbook/LiteLLM_HuggingFace.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

### Basic Completion

> Connect to your deployed Hugging Face Inference Endpoint for model inference. This provides dedicated infrastructure for your specific use case.

Reviewer 4: ```suggestion
After you have [deployed your Hugging Face Inference Endpoint](https://endpoints.huggingface.co/new) on dedicated infrastructure, you can run inference on it by providing the endpoint base URL in `api_base`, and indicating `huggingface/tgi` as the model name:
```

                    }
                },
            ],
        }
    ]
response = completion(
    model="huggingface/tgi",
    messages=messages,
    api_base="https://my-endpoint.endpoints.huggingface.cloud/v1/""
)
print(response.choices[0])
```

### Function Calling

```python
import os
from litellm import completion

os.environ["HF_TOKEN"] = "hf_xxxxxx"

functions = [{
    "name": "get_weather",
    "description": "Get the weather in a given location",
    "parameters": {
        "type": "object",
        "properties": {
            "location": {
                "type": "string",
                "description": "The location to get weather for"
            }
        },
        "required": ["location"]
    }
}]

response = completion(
    model="huggingface/tgi",
    messages=[{"content": "What's the weather like in San Francisco?", "role": "user"}],
    api_base="https://my-endpoint.endpoints.huggingface.cloud/v1/",
    functions=functions
)
print(response)
```

</TabItem>
</Tabs>

## LiteLLM Proxy Server with Hugging Face models

> Here is how to setup a [LiteLLM Proxy Server](https://docs.litellm.ai/#litellm-proxy-server-llm-gateway) to use Hugging Face models and how to use it:

Reviewer 4: ```suggestion
You can set up a [LiteLLM Proxy Server](https://docs.litellm.ai/#litellm-proxy-server-llm-gateway) to serve Hugging Face models through any of the supported Inference Providers. Here's how to do it:
```

> ### Step 1. Setup the config file

Reviewer 4: ```suggestion
### Step 1. Setup the config file

In this case, we are configuring a proxy to serve `DeepSeek R1` from Hugging Face, using Together AI as the backend Inference Provider.

```

```yaml
model_list:
  - model_name: my-r1-model
    litellm_params:
      model: huggingface/together/deepseek-ai/DeepSeek-R1
      api_key: os.environ/HF_TOKEN # ensure you have `HF_TOKEN` in your .env
```

### Step 2. Start the server
```bash
litellm --config /path/to/config.yaml
```

### Step 3. Make a request to the server
<Tabs>
<TabItem value="curl" label="curl">

```shell
curl --location 'http://0.0.0.0:4000/chat/completions' \
    --header 'Content-Type: application/json' \
    --data '{
    "model": "my-r1-model",
    "messages": [
        {
            "role": "user",
            "content": "Hello, how are you?"
        }
    ]
}'
```

</TabItem>
<TabItem value="python" label="python">

```python
# pip install huggingface-hub
from huggingface_hub import InferenceClient

client = InferenceClient(
    base_url="http://0.0.0.0:4000"
)

response = client.chat.completions.create(
    model="my-r1-model",
    messages=[
        {"role": "user", "content": "Hello, how are you?"}
    ]
)
print(response)
```

</TabItem>

> </Tabs>

Reviewer 4: ```suggestion
</Tabs>

For more information, please check the [Hugging Face Inference Client documentation](https://huggingface.co/docs/huggingface_hub/en/package_reference/inference_client). 
```

## Embedding

> LiteLLM supports Hugging Face's [text-embedding-inference](https://github.com/huggingface/text-embeddings-inference) models as well.

Reviewer 4: Does this work through inference providers as well, or do users have to use the HF model id? [The model below](https://huggingface.co/microsoft/codebert-base), for instance, does not seem to be available through the widget.

Author: no it's not, the users have to use the HF model id

```python
from litellm import embedding
import os
os.environ['HF_TOKEN'] = "hf_xxxxxx"
response = embedding(
    model='huggingface/microsoft/codebert-base',
    input=["good morning from litellm"]
)
```

# FAQ

**How does billing work with Hugging Face Inference Providers?**

> > You pay the standard provider API rates with no additional markup - Hugging Face simply passes through the provider costs. Note that [Hugging Face PRO](https://huggingface.co/subscribe/pro) users get $2 worth of Inference credits every month that can be used across providers.

Reviewer 2: ```suggestion
> Billing is centralized on your Hugging Face account, no matter which providers you are using. You are billed the standard provider API rates with no additional markup - Hugging Face simply passes through the provider costs. Note that [Hugging Face PRO](https://huggingface.co/subscribe/pro) users get $2 worth of Inference credits every month that can be used across providers.

**Do I need to create an account for each Inference Provider?**

> No, you don't need to create separate accounts. All requests are routed through Hugging Face, so you only need your HF token. This allows you to easily benchmark different providers and choose the one that best fits your needs.
```

I'd add a sentence to say that billing is centralized + an explicit question to showcase why it's nice to use HF routing (==no need for separate accounts)

![Filter models by Inference Provider](../../img/hf_filter_inference_providers.png)

### Dedicated Inference Endpoints
Refer to the [Inference Endpoints catalog](https://endpoints.huggingface.co/catalog) for a list of available models.

## Usage

<Tabs>
<TabItem value="serverless" label="Serverless Inference Providers">

### Authentication
With a single Hugging Face token, you can access inference through multiple providers. Your calls are routed through Hugging Face and the usage is billed directly to your Hugging Face account at the standard provider API rates.

Simply set the `HF_TOKEN` environment variable with your Hugging Face token, you can create one here: https://huggingface.co/settings/tokens.

```bash
export HF_TOKEN="hf_xxxxxx"
```
or alternatively, you can pass your Hugging Face token as a parameter:
```python
completion(..., api_key="hf_xxxxxx")
```

### Getting Started

To use a Hugging Face model, specify both the provider and model you want to use in the following format:
```

> huggingface/<provider>/<model_id>

Reviewer: ```suggestion
huggingface/<provider>/<hf_org_or_user>/<hf_model>
```

maybe like this to show the last part also has a `/` in it (aka., is namespaced)?

```

> Where `<model_id>` is the Hugging Face model ID and `<provider>` is the inference provider.

Reviewer: ```suggestion
Where `<hf_org_or_user>/<hf_model>` is the Hugging Face model ID and `<provider>` is the inference provider.
```
